R/estimate_uncertainty.R

```diff
@@ -0,0 +1,213 @@
+#'   triangle_to_nowcast = triangle,
+#'   delay_pmf = delay_pmf,
+#'   n_history_dispersion = 5
+#' )
+estimate_uncertainty <- function(triangle_to_nowcast,
+                                 delay_pmf,
+                                 n_history_dispersion = nrow(triangle_to_nowcast) - 1) { # nolint
+
+  n_horizons <- length(delay_pmf) - 1
+
+  # Add validation for n_history_dispersion, must be less than
+  # number of rows in the triangle minus one
+  .validate_triangle(triangle_to_nowcast)
+  .validate_delay_and_triangle(
+    triangle = triangle_to_nowcast,
+    delay_pmf = delay_pmf
+  )
+  if (n_history_dispersion > nrow(triangle_to_nowcast) - 1) {
+    cli_abort(
+      message = c(
+        "Triangle to nowcast does not contain sufficient rows to ",
+        "estimate uncertainty from `n_history_dispersion` observations. Either",
+        "pass in a triangle of more rows or lower the `n_history_dispersion`"
+      )
+    )
+  }
+
+  # Get the truncated matrix of observations you will use to estimate the
+  # dispersion (get rid of early rows that we're not using and add NAs to
+  # bottom right of the triangle)
+  matr_observed <- triangle_to_nowcast[
+    (nrow(triangle_to_nowcast) - n_history_dispersion):
+    nrow(triangle_to_nowcast),
+  ] |>
+    .replace_lower_right_with_NA()
+
+  results <- lapply(seq_len(n_history_dispersion), function(t) {
+    # Truncate the matrix observed by ignoring rows after t, replace rows that
+    # wouldn't be observed with NAs
+    matr_observed_temp <- matrix(matr_observed[1:t, ], nrow = t) |>
+      .replace_lower_right_with_NA()
+    # From the truncated would have been observed as of t matrix, get the
+    # expected nowcast as of t for each d
+    matr_exp_temp <- apply_delay(
+      triangle_to_nowcast = matr_observed_temp,
+      delay_pmf = delay_pmf
+    )
+
+    # We now have what we would have estimated for each delay,
+    # and the corresponding observations for those reference dates and delays
+    # (if they are present).
+    trunc_matr_observed <- matrix(matr_observed[1:t, ], nrow = t)
+    indices_nowcast <- matrix(is.na(matr_observed_temp), nrow = t)
+    indices_observed <- matrix(!is.na(trunc_matr_observed), nrow = t)
+
+    # First delay is 0, there will never be any nowcasts. We skip to
+    # delay = 1 (index of 2), and apply the logic for each horizon, finding the
+    # indices that were both nowcasted and observed, and estimating what we
+    # would have added at that delay vs what we observed.
+    exp_to_add <- sapply((seq_len(n_horizons) + 1),
+      .conditional_sum_cols,
+      matrix_bool1 = indices_nowcast,
+      matrix_bool2 = indices_observed,
+      matrix_to_sum = matr_exp_temp
+    )
+    to_add <- sapply((seq_len(n_horizons) + 1),
+      .conditional_sum_cols,
+      matrix_bool1 = indices_nowcast,
+      matrix_bool2 = indices_observed,
+      matrix_to_sum = trunc_matr_observed
+    )
+    return(
+      list(
+        exp_to_add = exp_to_add,
```

Comment:
it seems like you might have more fun returning a data.frame

Reply:
Yeah this is a bit clunky, will rework to return a data.frame

Reply:
Ok originally was thinking of just doing the rbind within the loop, but perplexity says thats not best practice. I had changed to have the function return a dataframe, but still do the rbind outside the loop

Reply:
yes I agree